### PR TITLE
Replaced usage of Ray with Vector #21

### PR DESCRIPTION
**Pull request by azoitl**
Draw2d's Ray class is deprecated. This commit removes all its usage and replaces it with the Vector class. To make migration easier two new constructors where added to Vector.

Addresses https://github.com/eclipse/gef-classic/issues/21

### CODE REVIEW

File: org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java [hunk 1]
```diff
@@ -361,20 +363,20 @@ public void route(Connection conn) {
...
-				positions.add(Integer.valueOf(i));
+				positions.add(Double.valueOf(i));
 				horizontal = !horizontal;
 
-				if (startNormal.dotProduct(direction) < 0) {
-					i = endNormal.similarity(end.getAdded(endNormal.getScaled(10)));
-					positions.add(Integer.valueOf(i));
+				if (startNormal.getDotProduct(direction) < 0) {
+					i = endNormal.getSimilarity(end.getAdded(endNormal.getMultiplied(10)));
+					positions.add(Double.valueOf(i));
```
Comment (ptziegler) [inline org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java:371]:
```suggestion
					positions.add(i);
```

File: org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java [hunk 2]
```diff
@@ -286,73 +288,73 @@ public void route(Connection conn) {
...
+					i = endNormal.getSimilarity(end.getAdded(endNormal.getMultiplied(10)));
 				} else {
 					if (horizontal) {
 						i = average.y;
 					} else {
 						i = average.x;
 					}
 				}
-				positions.add(Integer.valueOf(i));
+				positions.add(Double.valueOf(i));
```
Comment (ptziegler) [inline org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java:340]:
```suggestion
				positions.add(i);
```

**Review by ptziegler — COMMENTED** [review]
I'm a little bit concerned about the Vector introducing inaccuracies due to fractions, but I couldn't see anything wrong while testing. So I think it's fine to simply swap.

File: org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java [hunk 3]
```diff
@@ -286,73 +288,73 @@ public void route(Connection conn) {
...
+		Vector direction = new Vector(start, end);
+		Vector startNormal = getStartDirection(conn);
+		Vector endNormal = getEndDirection(conn);
 
-		List<Integer> positions = new ArrayList<>(5);
+		List<Double> positions = new ArrayList<>(5);
 		boolean horizontal = startNormal.isHorizontal();
 		if (horizontal) {
-			positions.add(Integer.valueOf(start.y));
+			positions.add(Double.valueOf(start.y));
```
[the remark this answers is not in the document]
Comment (azoitl) [reply org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java:308]: In general I would agree. However Sonar and the Java compiler is not happy therefore I added it.

[the remark this answers is not in the document]
Comment (ptziegler) [reply org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java:308]: Can you give an example what those warnings look like? I don't really see how auto-boxing primitives would be problematic, given that the Java compiler does effectively the same under the hood...

[the remark this answers is not in the document]
Comment (azoitl) [reply org.eclipse.draw2d/src/org/eclipse/draw2d/ManhattanConnectionRouter.java:308]: Currently my local sonarlint is not working. Not sure what I've messed up here. In the sonarqube server we currently do not have such a warning. Given the readability of the code I'll remove it and we will see.

File: org.eclipse.draw2d/src/org/eclipse/draw2d/geometry/Vector.java [hunk 4]
```diff
@@ -50,9 +50,20 @@ public Vector(PrecisionPoint p) {
...
+	 */
+	public Vector(Point p) {
+		x = p.preciseX();
+		y = p.preciseY();
+	}
+
 	/**
 	 * Constructs a Vector representing the direction and magnitude between to
-	 * provided Points.
+	 * provided PrecisioinPoints.
```
[the remark this answers is not in the document]
Comment (ptziegler) [reply org.eclipse.draw2d/src/org/eclipse/draw2d/geometry/Vector.java:66]: @azoitl There is still this minor typo here.

[the remark this answers is not in the document]
Comment (azoitl) [reply org.eclipse.draw2d/src/org/eclipse/draw2d/geometry/Vector.java]: Sorry missed that. Fixed now.

**Comment by azoitl** [comment]
It seems Eclipse Jenkins has issues accessing maven central again.

**Comment by ptziegler** [comment]
> It seems Eclipse Jenkins has issues accessing maven central again.

I think this issue lies with `repo.eclipse.org`, at least that's what I've noticed in the past. From what I understand, the `403 forbidden` error is a fancy way of saying that the repository is down.

https://www.eclipsestatus.io/incident/422280

> CI services, such as Jiro and Nexus (repo.eclipse.org) are in a degraded state, as the underlying storage device is incapable of handling peak loads. Work is underway to rectify the problem.

Hopefully, the repo was only briefly unavailable and the job will finish now.

**Comment by ptziegler** [comment]
Closes [Bug 310723](https://bugs.eclipse.org/bugs/show_bug.cgi?id=310723) - FanRouter and ManhattanConnectionRouter need to be adopted to replacement of Ray with Vector.

**Pull request merged**